Make interface compute items work with cache tags

nilsvu opened the pull request:
## Proposed changes

Tags from the ConstGlobalCache are now available through the DataBox,
but `db::item_type` didn't work for them. Since `Tags::InterfaceComputeItem`
needs `db::item_type`, this restriction prohibited cache tags from being used as
argument tags.

### Types of changes:

- [x] Bugfix

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsdeppe reviewed (changes requested):
I don't understand this change at all. Any `SimpleTag` has a `type` type alias, otherwise it's not a simple tag and you cannot retrieve the type easily. This is quite literally adding a bug into the code.

nilsvu commented:
Try to retrieve a tag added through `Tags::FromConstGlobalCache` in an interface compute item. It will give a compiler error because those tags don't derive from `SimpleTag`, so don't work with `db::item_type` right now.

nilsdeppe commented:
Yes, that's exactly what should happen.

nilsvu commented:
Those cache tags are available in normal compute items by adding them to `argument_tags`. They should also be available in interface compute items by adding them to `argument_tags` (and `volume_tags`). If this is unclear, let's clarify this in a hangout instead of in this chat.

nilsdeppe commented:
You *want* them to be available there, which is reasonable. But what you're changing here introduces a bug. I understand what you're saying, but you should not change `DataBox` because of an `Interface` deficiency.

nilsvu commented:
do you have a better suggestion? I don't feel like arguing in chat

nilsdeppe commented:
Make `FromConstGlobalCache` be both a simple tag and a compute tag, then set the `type` alias there to be `ConstGlobalCache_detail::type_for_get<CacheTag, Metavariables>`.

nilsdeppe commented:
Actually, probably won't do what you want. All option tags need to be simple DataBox tags and the `FromConstGlobalCache` needs to verify the type is correct. Ultimately, no I don't think what you want can work in general.

nilsdeppe commented:
The only thing I can think of is having a simple tag `FromConstGlobalCache` that computes the type correctly from the option tag. I don't think anything else will work.
1. Rename `FromConstGlobalCache` -> `FromConstGlobalCacheCompute`.
2. Add `FromConstGlobalCache` simple tag that computes the **correct** type from the option tag.
3. Use `FromConstGlobalCache<Blah>` on interfaces.

nilsvu commented:
or require option tags that should be available through the databox derive from `db::SimpleTag`?

nilsdeppe commented:
That won't work once #1582 is merged

nilsvu commented:
If `db::get<OptionTag>(box)` works than also `db::item_type<OptionTag>` should work...

nilsdeppe commented:
No, it shouldn't and doesn't... That's never been guaranteed, or it was like 2 years ago.

nilsvu commented:
That seems confusing for anyone writing code that uses the DataBox.. they shouldn't have to know where the tags in the DataBox come from to be able to use `db::` functions on them

nilsdeppe commented:
Well, it literally cannot be changed. The reason is you cannot get the type of an abstract base tag without knowing all the tags in the DataBox. :man_shrugging: Cannot be changed.

nilsvu commented:
We could introduce a `TypedTag` protocol/concept/etc that requires a struct has a `type`. DataBox tags, option tags and tagged tuple tags would conform to it

nilsdeppe commented:
The problem is that `option_tag::type` is not necessarily equal to the type the DataBox will hold, and for pure base tags they don't even need to have a nested type alias...

wthrowe commented:
I have not checked the current code, but it may be possible to eliminate the uses of `db::item_type` in the interface code by passing the types around as more template parameters.

nilsvu commented:
@nilsdeppe implemented your suggestion. Not sure what subtleties need to be handled to compute the type of the cache tag "correctly", so if I should make it anything else in this PR than the current `CacheTag::type` please let me know. Perhaps with #1582 this needs updating.

nilsdeppe commented on src/Parallel/ConstGlobalCache.hpp:
```diff
@@ -234,7 +234,20 @@ struct ConstGlobalCacheImpl : ConstGlobalCache, db::SimpleTag {
+  static std::string name() noexcept {
+    return "FromConstGlobalCache(" + pretty_type::short_name<CacheTag>() + ")";
+  }
+  using type = typename CacheTag::type;
```
This still doesn't work...

nilsvu replied:
see my comment above please

nilsvu replied:
https://github.com/sxs-collaboration/spectre/pull/1583#issuecomment-505876162

nilsdeppe commented:
I don't have time for this right now.

nilsdeppe commented:
An offline discussion with @kidder he presented (in words) the following solution (in code):

```cpp
template <class CacheTag>
struct FromConstGlobalCache : CacheTag::container_tag, db::ComputeTag {
  static_assert(
      std::is_base_of<db::SimpleTag, typename CacheTag::container_tag>::value,
      "The CacheTag::container_tag must be a simple DataBox tag.");
  static std::string name() noexcept {
    return "FromConstGlobalCache(" + pretty_type::short_name<CacheTag>() + ")";
  }
  template <class Metavariables>
  static const ConstGlobalCache_detail::type_for_get<CacheTag, Metavariables>&
  function(const Parallel::ConstGlobalCache<Metavariables>* const& cache) {
    static_assert(std::is_same<db::item_type<typename CacheTag::container_tag>,
                               ConstGlobalCache_detail::type_for_get<
                                   CacheTag, Metavariables>>::value,
                  "The type of the container tag of the option tag must match "
                  "the tag in the global cache.");
    return Parallel::get<CacheTag>(*cache);
  }
  using argument_tags = tmpl::list<ConstGlobalCache>;
};
```

This means that `OptionTags` added to the global cache need to have a `container_tag` type alias which points to a simple tag that is used to retrieve them. I think this is quite clean and also makes the code a bit easier to use conceptually.

nilsvu committed:
Add a container_tag to option tags that are available through the DataBox

This allows to use the container tag to retrieve the cache tag like any
other simple tag. In particular, it can be used in compute items.

nilsvu commented:
Thank you for the suggestion, I think this would be a very clean solution. I actually went through the code and implemented this for all the option tags that are used in the cache, and it works fine except for an issue with `OptionTags::TypedTimeStepper` which has a `type = std::unique_ptr<StepperType>`. The compiler complains that the types from the cache and the tag are not equal, even though both are `std::unique_ptr<StepperType>`. Any ideas @kidder @wthrowe @nilsdeppe ?

nilsdeppe commented:
@nilsleiffischer without a compiler error output it's nearly impossible to help. My guess is that the `unique_ptr` is derferenced and so that needs to be accounted for.

nilsvu commented:
@kidder @wthrowe @nilsdeppe This is fixed now. I'm not 100% sure if I handle cache tags with `std::unique_ptr`s correctly, but I didn't change the tests in `Unit.Parallel.ConstGlobalCacheDataBox` and they still pass, so that's good :) Please check this when you review.

nilsvu commented:
@nilsdeppe @kidder @wthrowe could you have a look at this when you get the chance?

wthrowe commented on src/IO/Observer/Tags.hpp:
```diff
@@ -131,6 +131,23 @@ struct H5FileLock : db::SimpleTag {
+///
+/// \see observers::OptionTags::VolumeFileName
+struct VolumeFileName : db::SimpleTag {
+  static std::string name() { return "ObserversVolumeFileName"; }
```
Why are the Tag and OptionTag `name`s different?  Particularly since I don't think the Tag names will be used for anything except observing and observing input file options is not very useful.

Maybe we should also add a default implementation of `name` for DataBox tags mirroring the one for option tags.

nilsvu replied:
I'm replying here also to your review message, since it's easier to keep track of comments in a thread.

1. I think the idea is that the option tag and the data box tag are conceptually different. The option tag defines a segment in the input file that yields an object. The databox tag gives an object meaning in the context of a parallel component. In particular, the databox tag should make sense on its own - it's just a descriptive label for an object. It _can_ be provided by an option tag, but doesn't have to be. The best example is probably `Tags::Domain`, which is commonly provided by `OptionTags::DomainCreator` but doesn't have to be. I think it's the cleanest approach to make this distinction for all option tags. In fact, while doing the refactoring I appreciated that I could choose names and such in the context of options and in the context of the databox separately.

    Please also consider this: In some cases the option tag carries a `std::unique_ptr`, e.g. `OptionTags::TimeStepper`. The databox tag `Tags::TimeStepper` can omit the pointer, since it will be provided by a derived compute item that returns a const ref. This again shows the distinction between the option tag providing the object and the databox tag labelling it in the context of the parallel component.

    Most importantly, on retrieval of a tag the user should not have to care how the tag got into the databox, just that it's there.

2. I don't feel strongly about the name `container_tag`, we could make it `databox_tag` if you prefer? Please make a suggestion.

3. Some `name`s differ because I tried to make them unique in the databox by including some variation of their namespace as a prefix. That's because I think the name is mostly to retrieve a tag by name at runtime. I don't feel strongly about that, either.

I'd appreciate if you could reach a conclusion on these matters with @nilsdeppe and @kidder at Cornell and let me know which changes to make, to make this review more efficient.

nilsvu replied:
Another suggestion: we could derive the option tags from their respective simple tag. Then we don't need a `container_tag` alias.

wthrowe commented on tests/Unit/Parallel/Test_ConstGlobalCache.cpp:
```diff
@@ -88,14 +94,16 @@ class Square : public Shape {
 struct shape_of_nametag_base {};
 
-struct shape_of_nametag : shape_of_nametag_base {
+struct shape_of_nametag : shape_of_nametag_base, db::SimpleTag {
```
If you're going to fix the names on some of the classes in this file you might as well do them all.

nilsvu replied:
I only had to fix the `name`->`Name` tag to avoid collision with its `name()` function. But sure, I can change the other names, too.

wthrowe reviewed (changes requested):
Why is this option called `container_tag`?  It doesn't seem to have anything to do with containers.

It seems like in almost all cases the option tag and the container tag are identical except for a help string (and the new `container_tag`).  Can we reuse them or even default `container_tag` to the same tag if it is not present?

nilsdeppe commented:
@wthrowe @kidder and I came up with the following interface on Wednesday:

```cpp
namespace OptionTags {
template <size_t Dim, typename TargetFrame>
struct DomainCreator {
  using type = std::unique_ptr<::DomainCreator<Dim, TargetFrame>>;
  static constexpr OptionString help = {"The domain to create initially"};
};

}  // namespace OptionTags

namespace Tags {
template <size_t VolumeDim, typename Frame>
struct Domain : db::SimpleTag {
  static std::string name() noexcept { return "Domain"; }
  using type = ::Domain<VolumeDim, Frame>;

  using option_tags = tmpl::list<::OptionTags::DomainCreator<VolumeDim, Frame>>;
  static ::Domain<VolumeDim, Frame> convert_from_options(
      const std::unique_ptr<::DomainCreator<VolumeDim, Frame>>&
          option_data) noexcept {
    return option_data->create_domain();
  }
};

template <size_t VolumeDim, typename Frame>
struct InitialExtents : db::SimpleTag {
  static std::string name() noexcept { return "Domain"; }
  using type = std::array<...>;

  using option_tags = tmpl::list<::OptionTags::DomainCreator<VolumeDim, Frame>>;
  static std::array<...> convert_from_options(
      const std::unique_ptr<::DomainCreator<VolumeDim, Frame>>&
          option_data) noexcept {
    return option_data->initial_extents();
  }
};


template <typename ParallelComponent>
struct Component : db::SimpleTag, ParallelComponent::pure_base_tag {
  using type = ParallelComponent::cproxy;
};
}  // namespace Tags
```

`std::unique_ptr`s will need to be handled with some care, but won't be an actual problem.

nilsvu commented:
Labeling `in progress` because @kidder is implementing this

nilsdeppe commented:
@nilsleiffischer can this be closed since #1666 is now merged?

nilsvu commented:
Yes 👍
